src/dhtproto/node/neo/request/Update.d

```diff
@@ -0,0 +1,329 @@
+        (*buf).copy(init_payload);
+        this.init_payload = *buf;
+    }
+
+    /***************************************************************************
+
+        Called by the connection handler after the supported code has been sent
+        back to the client.
+
+    ***************************************************************************/
+
+    public void postSupportedCodeSent ( )
+    {
+        auto ed = this.connection.event_dispatcher();
+
+        auto message =
+            *ed.message_parser.getValue!(MessageType)(this.init_payload);
+        auto channel = ed.message_parser.getArray!(char)(this.init_payload);
+        auto key = *ed.message_parser.getValue!(hash_t)(this.init_payload);
+
+        MessageType response;
+
+        if ( this.responsibleForKey(key) )
+        {
+            with ( MessageType ) switch ( message )
+            {
+                // Normal sequence: read, wait for response, update.
+                case GetRecord:
+                    response = this.getWaitUpdate(channel, key);
+                    break;
+
+                // Either an updated record transferred from another node or a
+                // new record that did not exist in the DHT before. Just put it.
+                case UpdateRecord:
+                    response = this.update(channel, key, this.init_payload);
+                    break;
+
+                default:
+                    ed.shutdownWithProtocolError("Unexpected message from client");
+            }
+        }
+        else
+            response = MessageType.WrongNode;
+
+        // Send the final message to the client.
+        ed.send(
+            ( ed.Payload payload )
+            {
+                payload.add(response);
+            }
+        );
+        ed.flush();
+    }
+
+    /***************************************************************************
+
+        Normal sequence: get, wait for response, update.
+
+        Params:
+            channel = channel to update record in
+            key = key of record to update
+
+        Returns:
+            message type code to return to the client
+
+    ***************************************************************************/
+
+    private MessageType getWaitUpdate ( cstring channel, hash_t key )
+    {
+        auto ed = this.connection.event_dispatcher();
+
+        // Get record value from storage and send it to the client.
+        bool exists;
+        auto success = this.get(channel, key,
+            ( Const!(void)[] value )
+            {
+                exists = true;
+
+                ed.send(
+                    ( ed.Payload payload )
+                    {
+                        payload.addCopy(MessageType.RecordValue);
+                    }
+                );
+                ed.flush();
+
+                // TODO: The value could be sent in the same message as the
+                // response. For simplicity, we send them separately, at
+                // present.
```

Comment:
Why is this simpler?

Reply:
It saves having to interpret one message in different ways, on the client side, depending on the contents.